server/packetio.go

```diff
@@ -80,12 +93,48 @@ func (p *packetIO) setReadTimeout(timeout time.Duration) {
 
 func (p *packetIO) readOnePacket() ([]byte, error) {
 	var header [4]byte
+	r := io.NopCloser(p.bufReadConn)
 	if p.readTimeout > 0 {
 		if err := p.bufReadConn.SetReadDeadline(time.Now().Add(p.readTimeout)); err != nil {
 			return nil, err
 		}
 	}
-	if _, err := io.ReadFull(p.bufReadConn, header[:]); err != nil {
+	if p.compressionAlgorithm != mysql.CompressionNone {
+		var compressedHeader [7]byte
+		if _, err := io.ReadFull(p.bufReadConn, compressedHeader[:]); err != nil {
+			return nil, errors.Trace(err)
+		}
+		compressedSequence := compressedHeader[3]
+		if compressedSequence != p.compressedSequence {
+			return nil, errInvalidSequence.GenWithStack(
+				"invalid compressed sequence %d != %d", compressedSequence, p.compressedSequence)
+		}
+		p.compressedSequence++
+		p.compressedWriter.compressedSequence = p.compressedSequence
+		uncompressedLength := int(uint32(compressedHeader[4]) | uint32(compressedHeader[5])<<8 | uint32(compressedHeader[6])<<16)
+
+		if uncompressedLength > 0 {
+			var zr io.ReadCloser
+			switch p.compressionAlgorithm {
+			case mysql.CompressionZlib:
+				var err error
+				zr, err = zlib.NewReader(p.bufReadConn)
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+			case mysql.CompressionZstd:
+				zstdReader, err := zstd.NewReader(p.bufReadConn, zstd.WithDecoderConcurrency(1))
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+				zr = zstdReader.IOReadCloser()
+			default:
+				return nil, errors.New("Unknown compression algorithm")
+			}
+			r = zr
```

Comment:
can we directly assign to `r`?

Reply:
Yes: 6484a7efb1e7630d1c2ab364e0ff54da2fa788ae